tidb_binlog/pump_client/client.go

```diff
@@ -0,0 +1,393 @@
+	DefaultHeartbeatWaitTime = 30 * time.Second
+
+	aliveStr  = "alive"
+	objectStr = "object"
+
+	fakeClusterID uint64 = 110119120
+)
+
+// PumpsClient is the client of pumps.
+type PumpsClient struct {
+	sync.RWMutex
+
+	ctx context.Context
+
+	cancel context.CancelFunc
+
+	wg sync.WaitGroup
+
+	// the client of etcd.
+	EtcdCli *etcd.Client
+
+	// Pumps saves the whole pumps' status.
+	Pumps []*PumpStatus
+
+	// PumpMap saves the map of pump's node and pump status.
+	PumpMap map[string]*PumpStatus
+
+	// AvliablePumps saves the whole avaliable pumps' status.
+	AvaliablePumps []*PumpStatus
+
+	// NeedCheckPumps saves the pumps need to be checked.
+	NeedCheckPumps []*PumpStatus
+
+	// Selector will select a suitable pump.
+	Selector PumpSelector
+
+	// the max retry time if write binlog failed.
+	RetryTime int
+
+	// BinlogWriteTimeout is the max time binlog can use to write to pump.
+	BinlogWriteTimeout time.Duration
+}
+
+// NewPumpsClient returns a PumpsClient.
+func NewPumpsClient(etcdURLs string, security *tls.Config, algorithm string) (*PumpsClient, error) {
+	var selector PumpSelector
+	switch algorithm {
+	case Hash:
+		selector = NewHashSelector()
+	case Score:
+		selector = NewScoreSelector()
+	default:
+		log.Warnf("unknow algorithm %s, use hash as default", algorithm)
+		selector = NewHashSelector()
+	}
+
+	ectdEndpoints, err := utils.ParseHostPortAddr(etcdURLs)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	cli, err := etcd.NewClientFromCfg(ectdEndpoints, DefaultEtcdTimeout, RootPath, security)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	newPumpsClient := &PumpsClient{
+		ctx:                ctx,
+		cancel:             cancel,
+		EtcdCli:            cli,
+		Selector:           selector,
+		RetryTime:          DefaultRetryTime,
+		BinlogWriteTimeout: DefaultBinlogWriteTimeout,
+	}
+
+	err = newPumpsClient.getPumpStatus(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	newPumpsClient.Selector.SetPumps(newPumpsClient.AvaliablePumps)
+
+	newPumpsClient.wg.Add(2)
+	go newPumpsClient.watchStatus()
+	go newPumpsClient.heartbeat()
+
+	return newPumpsClient, nil
+}
+
+// getPumpStatus retruns all the pumps status in the etcd.
+func (c *PumpsClient) getPumpStatus(pctx context.Context) error {
+	c.Lock()
+	defer c.Unlock()
+
+	ctx, cancel := context.WithTimeout(pctx, DefaultEtcdTimeout)
+	defer cancel()
+
+	resp, err := c.EtcdCli.List(ctx, path.Join(RootPath))
+	if err != nil {
+		return errors.Trace(err)
+	}
+	nodesStatus, err := node.NodesStatusFromEtcdNode(resp)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for _, status := range nodesStatus {
+		log.Infof("get pump %v from etcd", status)
+		c.addPump(status)
+	}
+
+	return nil
+}
+
+// WriteBinlog writes binlog to a situable pump.
+func (c *PumpsClient) WriteBinlog(clusterID uint64, binlog *pb.Binlog) error {
+	c.RLock()
+	pump := c.Selector.Select(binlog)
```

Comment:
is Select() of Selector supposed to be thread-safe? i see  Write Lock inside the whole Select() func, and why here need a RLock in PumpsClient?